crates/core/src/execution_context.rs

```diff
@@ -1,17 +1,125 @@
+use std::cell::RefCell;
+
 use derive_more::Display;
 use spacetimedb_lib::Address;
+use spacetimedb_primitives::TableId;
+
+use crate::db::db_metrics::DB_METRICS;
+
+pub enum MetricType {
+    RdbNumIndexSeeks,
+    RdbNumKeysScanned,
+    RdbNumRowsFetched,
+}
+#[derive(Default, Clone)]
+struct BufferMetric {
+    pub table_id: TableId,
+    pub rdb_num_index_seeks: u64,
+    pub rdb_num_keys_scanned: u64,
+    pub rdb_num_rows_fetched: u64,
+    pub cache_table_name: String,
+}
+
+impl BufferMetric {
+    pub fn inc_by(&mut self, ty: MetricType, val: u64) {
+        match ty {
+            MetricType::RdbNumIndexSeeks => {
+                self.rdb_num_index_seeks += val;
+            }
+            MetricType::RdbNumKeysScanned => {
+                self.rdb_num_keys_scanned += val;
+            }
+            MetricType::RdbNumRowsFetched => {
+                self.rdb_num_rows_fetched += val;
+            }
+        }
+    }
+}
+
+impl BufferMetric {
+    pub fn new(table_id: TableId, table_name: &str) -> Self {
+        Self {
+            table_id,
+            rdb_num_index_seeks: 0,
+            rdb_num_keys_scanned: 0,
+            rdb_num_rows_fetched: 0,
+            cache_table_name: table_name.to_string(),
```

Comment:
Would it be relatively simple to take a reference here rather than clone?

Reply:
Lifetime issues can be pain there, had to make sure `ctx` always outlive `table_name` will anaylyze this bit more.